### ISSUE TITLE
Unable to Apply Kinesis Stream Retention Change

### ISSUE THREAD

**Issue by AlexThurston** [hashicorp/terraform#7239]
### Affected Resource(s)
Please list the resources as a list, for example:
- kinesis

### Expected Behaviour
I would have expected the change in retention to be listed as a change and not an add and delete during the planning phase.

### Actual Behaviour
During the planning phase, a change to a kinesis stream's retention was listed as an add and delete (which I think was strange since I would have expected it to be a change).  When the change was applied, I receive this message to the console:

```
Error applying plan:

1 error(s) occurred:

* aws_kinesis_stream.ion_stream: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.6.16
    Resource ID: aws_kinesis_stream.ion_stream
    Mismatch reason: diff: Destroy; old: false, new: true
    Diff One (usually from plan): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff{"retention_period":*terraform.ResourceAttrDiff{Old:"24", New:"168", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}}, Destroy:false, DestroyTainted:false}
    Diff Two (usually from apply): *terraform.InstanceDiff{Attributes:map[string]*terraform.ResourceAttrDiff{"arn":*terraform.ResourceAttrDiff{Old:"arn:aws:kinesis:us-east-1:188813637777:stream/staging_cud", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "tags.#":*terraform.ResourceAttrDiff{Old:"1", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "tags.deployment":*terraform.ResourceAttrDiff{Old:"ion-staging", New:"ion-staging", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "name":*terraform.ResourceAttrDiff{Old:"staging_cud", New:"staging_cud", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}, "retention_period":*terraform.ResourceAttrDiff{Old:"24", New:"168", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Type:0x0}}, Destroy:true, DestroyTainted:false}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.


Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. Apply a change where the retention of a kinesis stream is changed.

**Comment by stack72** [comment]
Hi @AlexThurston 

Please can you post what the config was pre-update? I want to try and replicate this exactly

Thanks

Paul

**Comment by AlexThurston** [comment]
Hey @stack72,

The original resource was this:
```
# Kinesis Stream resources
resource "aws_kinesis_stream" "foo_stream" {
  name = "FooStream"
  shard_count = 1
  lifecycle {
    ignore_changes = ["shard_count"]
  }
}
```

Which became:
```
# Kinesis Stream resources
resource "aws_kinesis_stream" "foo_stream" {
  name = "FooStream"
  shard_count = 2
  lifecycle {
    ignore_changes = ["shard_count"]
  }
}
```

I really only am changing the `shard_count`.

I'm not totally sure this is what you're looking so let me know.

Cheers,
Alex

**Comment by stack72** [comment]
Hi @AlexThurston 

So `shard_count` is set as ForceNew - any changes in ShardCount will force a destroy / create action

Therefore, I guess this is why you are using the lifecycle {} block

The issue suggestions retention_count - is this not the case?

P.

**Comment by AlexThurston** [comment]
Oh man, really sorry.  I muddled up the issues and confused myself.  The issue is NOT with `shard-count` but with changing the `retention_period`.

In this case, the pre-state would have been as I wrote originally so the default retention of 24 hours would be chosen.  However, when I try to change the resource to:
```
# Kinesis Stream resources
resource "aws_kinesis_stream" "foo_stream" {
  name = "FooStream"
  shard_count = 1
  retention_period = 168
  lifecycle {
    ignore_changes = ["shard_count"]
  }
}
```

**Comment by AlexThurston** [comment]
And yes, the `lifecycle` is there so that we can split shards without recreating the stream.

**Comment by stack72** [comment]
Hi @AlexThurston 

Ok, so I have been trying to test this out now (against master) and here are the results:

```
resource "aws_kinesis_stream" "foo_stream" {
  name = "FooStream"
  shard_count = 1
  lifecycle {
    ignore_changes = ["shard_count"]
  }
}
```

```
% terraform apply
aws_kinesis_stream.foo_stream: Creating...
  arn:              "" => "<computed>"
  name:             "" => "FooStream"
  retention_period: "" => "24"
  shard_count:      "" => "1"
aws_kinesis_stream.foo_stream: Still creating... (10s elapsed)
aws_kinesis_stream.foo_stream: Still creating... (20s elapsed)
aws_kinesis_stream.foo_stream: Still creating... (30s elapsed)
aws_kinesis_stream.foo_stream: Still creating... (40s elapsed)
aws_kinesis_stream.foo_stream: Still creating... (50s elapsed)
aws_kinesis_stream.foo_stream: Creation complete

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

Created as expected. When I run another plan, I see the following:

```
% terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but
will not be persisted to local or remote state storage.

aws_kinesis_stream.foo_stream: Refreshing state... (ID: arn:aws:kinesis:us-west-2:881237884953:stream/FooStream)

No changes. Infrastructure is up-to-date. This means that Terraform
could not detect any differences between your configuration and
the real physical resources that exist. As a result, Terraform
doesn't need to do anything.
```

So, I changed the config to look as follows:

```
resource "aws_kinesis_stream" "foo_stream" {
  name = "FooStream"
  shard_count = 1
  retention_period = 168
  lifecycle {
    ignore_changes = ["shard_count"]
  }
}
```

The Terraform plan looks as follows:

```
% terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but
will not be persisted to local or remote state storage.

aws_kinesis_stream.foo_stream: Refreshing state... (ID: arn:aws:kinesis:us-west-2:881237884953:stream/FooStream)

The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed. Cyan entries are data sources to be read.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

~ aws_kinesis_stream.foo_stream
    retention_period: "24" => "168"


Plan: 0 to add, 1 to change, 0 to destroy.
```

The apply then gave me this:

```
% terraform apply
aws_kinesis_stream.foo_stream: Refreshing state... (ID: arn:aws:kinesis:us-west-2:881237884953:stream/FooStream)
aws_kinesis_stream.foo_stream: Modifying...
  retention_period: "24" => "168"
aws_kinesis_stream.foo_stream: Still modifying... (10s elapsed)
aws_kinesis_stream.foo_stream: Still modifying... (20s elapsed)
aws_kinesis_stream.foo_stream: Still modifying... (30s elapsed)
aws_kinesis_stream.foo_stream: Modifications complete

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.
```

So the config rolled out as expected, then 1 more plan gave me this:

```
% terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but
will not be persisted to local or remote state storage.

aws_kinesis_stream.foo_stream: Refreshing state... (ID: arn:aws:kinesis:us-west-2:881237884953:stream/FooStream)

No changes. Infrastructure is up-to-date. This means that Terraform
could not detect any differences between your configuration and
the real physical resources that exist. As a result, Terraform
doesn't need to do anything.
```

Therefore, I believe this issue has been fixed as a result of the state changes in 0.7 :)

Hope this is ok?

Paul

**Comment by AlexThurston** [comment]
Fair enough.  I'll pull down the 0.7 pre-release code and try it out against out deployment to see the outcome.

**Comment by stack72** [comment]
Hi @AlexThurston 

Please read the CHANGELOG (https://github.com/hashicorp/terraform/blob/master/CHANGELOG.md#07-unreleased) to make sure you understand the potential issues of upgrading to 0.7

Paul

**Comment by stack72** [comment]
Hi @AlexThurston 

Going to close this for now - please do add a comment if your issue is not resolved by 0.7 

Thanks

Paul
